harness/determined/_execution.py

```diff
@@ -16,6 +16,18 @@ def _get_gpus() -> Tuple[bool, List[str], List[int]]:
     return use_gpu, gpu_uuids, gpu_ids
 
 
+def _execute_user_function(func: Callable, *args: Any, **kwargs: Any) -> Any:
+    try:
+        res = func(*args, **kwargs)
+        return res
+    except SystemExit as e:
+        logging.warning(
+            "User code raised a SystemExit (sys.exit()) before entering "
+            "the training loop. Please remove this sys.exit() from your script."
+        )
+        raise e
```

Comment:
Can we do:
```python
    raise AssertionError("user code ... your script.") from e
```
instead?

Reply:
No, we can't. You raise `AssertionError` rather than the `SystemExit`. What we want is to just throw a warning to make users change their code.

Reply:
Ok, what is the desired system behavior?  In the jira ticket you reported that the trial would "fail silently".  If it is actually "failing" then we should not be exiting 0, which is an indicator that the process exited successfully.

Reply:
yes, that's fair. The use case in the ticket is that people use `tf.flags`, which exits with non-zero code if it hits errors. It might exit with a zero code, which I haven't confirmed. But anyways the desired behavior is that we should exit with non-zero code no matter which code users exit with and we shouldn't just "pass" users' exiting. `AssertionError` might be fine since we are going to exit anyway.

Reply:
Yes, also if we require users to fix it, exposing a stack trace is a nice thing to do

Reply:
When you `raise ..`, a stack trace is printed.

Reply:
I think the `SystemExit` may be special and it may not print a stack trace, which is why I suggested raising a different type of error.

Reply:
I tested it. It prints out stack trace of you use `raise .. from`. I wrote the following program:
```
import sys

def op():
    sys.exit(1)


print("before running sys.exit")
try:
    op()
except SystemExit as e:
    print("Caught SystemExit")
    raise AssertionError("re-raise error") from e
print("after running sys.exit")
```

What it prints out:
```
(determined) Shiyuans-MacBook-Pro-2:playground shiyuanzhu$ python sys_exit.py 
before running sys.exit
Caught SystemExit
Traceback (most recent call last):
  File "sys_exit.py", line 9, in <module>
    op()
  File "sys_exit.py", line 4, in op
    sys.exit(1)
SystemExit: 1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "sys_exit.py", line 12, in <module>
    raise AssertionError("re-raise error") from e
AssertionError: re-raise error
```